lib/services/role.go

```diff
@@ -1758,6 +1514,85 @@ func (set RoleSet) CheckLoginDuration(ttl time.Duration) ([]string, error) {
 	return out, nil
 }
 
+// CheckAccessToRemoteCluster checks if a role has access to remote cluster. Deny rules are
+// checked first then allow rules. Access to a cluster is determined by
+// namespaces, labels, and logins.
+//
+// Note, logging in this function only happens in debug mode, this is because
+// adding logging to this function (which is called on every server returned
+// by GetNodes) can slow down this function by 50x for large clusters!
+func (set RoleSet) CheckAccessToRemoteCluster(rc RemoteCluster) error {
+	var errs []error
+
+	rcLabels := rc.GetMetadata().Labels
+
+	// If there is no role in the set with labels and the cluster
+	// is not labelled ,assume that the role set has access to the cluster.
+	usesLabels := false
+	for _, role := range set {
+		if len(role.GetClusterLabels(Allow)) != 0 || len(role.GetClusterLabels(Deny)) != 0 {
+			usesLabels = true
+			break
+		}
+	}
+
+	if usesLabels == false && len(rcLabels) == 0 {
+		if log.GetLevel() == log.DebugLevel {
+			log.WithFields(log.Fields{
+				trace.Component: teleport.ComponentRBAC,
+			}).Debugf("Grant access to cluster %v - no role in %v uses cluster labels and the cluster is not labeled.",
+				rc.GetName(), set.RoleNames())
+		}
+		return nil
+	}
+
+	// Check deny rules first: a single matching namespace, label, or login from
+	// the deny role set prohibits access.
+	for _, role := range set {
+		matchLabels, labelsMessage, err := MatchLabels(role.GetClusterLabels(Deny), rcLabels)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		if matchLabels == true {
+			if log.GetLevel() == log.DebugLevel {
```

Comment:
this check shouldn't be necessary, since you call `log.Debugf`

Reply:
this is to avoid formatting call on a large scale

Reply:
https://github.com/sirupsen/logrus/blob/d131c24e23baaa812461202af6d7cfa388e2d292/entry.go#L328-L330 looks like the formatting call doesn't happen when log level check fails

Reply:
I still think this check is redundant (throughout this method)

Reply:
https://github.com/gravitational/teleport/issues/4681